### PR TITLE
Error reporting improvements

### PR DESCRIPTION
**Pull request by matei-p**

### CODE REVIEW

File: arwen/contexts/runtime.go [hunk 1]
```diff
@@ -917,6 +920,21 @@ func (context *runtimeContext) MemStore(offset int32, data []byte) error {
 	return nil
 }
 
+func (context *runtimeContext) AddError(err error, otherInfo ...string) {
+	if err == nil {
+		return
+	}
+	if context.errors == nil {
+		context.errors = arwen.WrapError(err, otherInfo...)
+	} else {
```
Comment (bogdan-rosianu) [inline arwen/contexts/runtime.go:929]: could do an early exit here.

```GOLANG
if condition {
    // do stuff
    return
}
// do stuff from the else branch
```

File: arwen/contexts/runtime.go [hunk 2]
```diff
@@ -917,6 +920,21 @@ func (context *runtimeContext) MemStore(offset int32, data []byte) error {
 	return nil
 }
 
+func (context *runtimeContext) AddError(err error, otherInfo ...string) {
```
Comment (bogdan-rosianu) [inline arwen/contexts/runtime.go:923]: maybe add comments to the exported functions

File: arwen/contexts/runtime.go [hunk 3]
```diff
@@ -917,6 +920,21 @@ func (context *runtimeContext) MemStore(offset int32, data []byte) error {
...
+	}
+	if context.errors == nil {
+		context.errors = arwen.WrapError(err, otherInfo...)
+	} else {
+		context.errors = context.errors.WrapWithError(err, otherInfo...)
+	}
+}
+
+func (context *runtimeContext) GetAllErrors() error {
+	return context.errors
```
Comment (bogdan-rosianu) [inline arwen/contexts/runtime.go:935]: Just a question: could we have concurrency issues here? Or all the times `GetAllErrors()` will be called only after all the errors will be added?

File: arwen/errorWrapping.go [hunk 4]
```diff
@@ -73,23 +76,41 @@ func (werr *wrappableError) GetLastError() error {
...
+func (werr *wrappableError) wrapWithErrorWithSkipLevels(err error, skipStackLevels int, otherInfo ...string) *wrappableError {
 	newErrs := make([]errorWithLocation, len(werr.errsWithLocation))
 	copy(newErrs, werr.errsWithLocation)
 	if err == nil {
 		return &wrappableError{
 			errsWithLocation: newErrs,
 		}
 	}
+
+	var errWithLocation []errorWithLocation
```
Comment (bogdan-rosianu) [inline arwen/errorWrapping.go:98]: `errsWithLocation` ?

File: go.sum [hunk 5]
```diff
@@ -65,8 +65,13 @@ github.com/ElrondNetwork/elrond-go v1.1.38-0.20210322140841-6a27c533e8a5/go.mod
 github.com/ElrondNetwork/elrond-go v1.1.42/go.mod h1:mxhdRRV7PY+IHzznJ3MnX5/2YSMNxl2znD6f6Y8nuCM=
 github.com/ElrondNetwork/elrond-go v1.1.51/go.mod h1:0gT8uLRXSqmKDrrn55aXTSca0XGpg8aszx+Q5OPMNpY=
 github.com/ElrondNetwork/elrond-go v1.1.59-0.20210526130950-2a93e2e11c39/go.mod h1:0aZfahOLX4vO+1D6Ejx8XwOLTzd7pfcx94xzAOpt2bs=
+<<<<<<< HEAD
```
Comment (bogdan-rosianu) [inline go.sum:68]: please finish the merge

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: arwen/contexts/runtime.go (hunk 3 above)
Comment (matei-p) [reply arwen/contexts/runtime.go:935]: yes, that should be the case, after an SC call

File: arwen/host/arwen.go [hunk 6]
```diff
@@ -431,3 +431,10 @@ func (host *vmHost) IsInterfaceNil() bool {
 func (host *vmHost) SetRuntimeContext(runtime arwen.RuntimeContext) {
 	host.runtimeContext = runtime
 }
+
+func (host *vmHost) GetRuntimeErrors() error {
```
Comment (sasurobert) [inline arwen/host/arwen.go:435]: add comment to exported function

File: arwen/contexts/runtime.go [hunk 7]
```diff
@@ -61,6 +63,7 @@ func NewRuntimeContext(host arwen.VMHost, vmType []byte, useWarmInstance bool) (
 		useWarmInstance:     useWarmInstance,
 		warmInstanceAddress: nil,
 		warmInstance:        nil,
+		errors:              nil,
```
Comment (sasurobert) [inline arwen/contexts/runtime.go:66]: clean up the errors on InitState

**Review by sasurobert — CHANGES_REQUESTED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**